middleware/cache/freq/freq.go

```diff
@@ -0,0 +1,39 @@
+// Package freq keeps track of last X seen events. The events themselves are not stored
+// here. So the Freq type should be added next to the thing it is tracking.
+package freq
+
+import "time"
+
+type Freq struct {
+	// Last time we saw a query for this element.
+	last time.Time
+	// Number of this in the last time slice.
+	hits int
+}
+
+func NewFreq(t time.Time) *Freq {
+	return &Freq{last: t, hits: 0}
+}
+
+// Updates updates the number of hits. Last time seen will be set to now.
+// If the last time we've seen this entity is within now - d, we increment hits, otherwise
+// we reset hits to 1. It returns the number of hits.
+func (f *Freq) Update(d time.Duration, now time.Time) int {
+	earliest := now.Add(-1 * d)
+	if f.last.Before(earliest) {
+		f.last = now
+		f.hits = 1
+		return f.hits
+	}
+	f.hits++
```

Comment:
Actually what I was thinking is we always update "last" too. Right now, this will build up for 1 minute, then go back to 1, then build up, then go back to 1. You'll only prefetch if you hit the threshold in the minute right before you check for purging it.

Instead, if you update last every time, it means there have been no gaps between queries that are as long a **DURATION**, and that this has persisted for at least **AMOUNT** queries. It's not the same measure as before, but to me at least seems like a valid measure of "popularity".